### PR TITLE
Vectorize all per-instance batch transforms

### PR DESCRIPTION
**Pull request by fepegar**
[Generated by a coding agent]

---

Stacked on #1473 (base: `per-instance-augmentation`).

## Vectorize all per-instance batch transforms

Per the principle that **batch transforms should always be vectorized**, this PR replaces the per-element `for index in range(B): … torch.cat` loops in the per-instance transforms with whole-batch tensor operations using broadcast per-element parameters.

### Transforms vectorized
- **Flip**: per-axis whole-batch flips selected with `torch.where`.
- **Ghosting / Spike**: a single batched `fftn`/`ifftn` with a per-element k-space mask (Ghosting) / per-element peak scatter (Spike).
- **BiasField**: stack per-seed coarse noise, then one batched `interpolate` → `exp` → multiply.
- **Blur**: grouped `conv3d` with per-element separable kernels (`groups = B*C`).
- **Motion**: batched FFTs + grid-resampled rigid transforms, mixing k-space segments across the batch.
- **LabelsToImage**: per-label batched generation with broadcast per-element stats.
- **Anisotropy**: group by spatial axis, batch the resample per group.
- **Swap**: batched indexed patch swaps.

### Correctness gate
`tests/conftest.py::assert_vectorized` (used by `tests/test_vectorization.py`) asserts that, for transforms whose `apply_transform` is deterministic given the recorded params, the batched result for each element equals applying that element's own parameters alone — proving the whole-batch computation is faithful per-element with no cross-element contamination. It also asserts gated-out elements are **bit-for-bit** no-ops. LabelsToImage (which samples inside `apply`) has a dedicated per-element correctness test.

Every transform preserves input dtype and leaves per-element gated-out elements (`_keep == False`) exactly unchanged. Motion was verified to match the previous implementation to float32 precision.

Full test suite green (1379 passed), `mise run quality` (ruff + ty) and `prek` (incl. Xenon complexity) green. A GPT 5.5 review was addressed (gate hardened to bit-exact gated rows; LabelsToImage per-element test added).

### CODE REVIEW

File: src/torchio/transforms/intensity/spike.py [hunk 1]
```diff
@@ -44,6 +44,8 @@ class Spike(IntensityTransform):
         >>> transform = tio.Spike(num_spikes=3, intensity=2.0)
     """
 
+    __name__ = "Spike"
```
[the remark this answers is not in the document]
Comment (fepegar) [reply src/torchio/transforms/intensity/spike.py]: [Generated by a coding agent]

---

Fixed in 32d60423. Removed the redundant `__name__ = "Spike"` class attribute (a leftover test-selection hack); `type` already provides `Spike.__name__`.

File: src/torchio/transforms/intensity/motion.py [hunk 2]
```diff
@@ -143,120 +149,405 @@ def _apply_motion(
...
+    translation: Tensor,
+) -> Tensor:
+    """Apply per-element rigid-body transforms to a 5-D tensor.
+
+    Each batch element gets its own affine grid, shared by all channels.
+
+    Args:
+        tensor: `(B, C, I, J, K)` tensor.
+        degrees: Euler angles in degrees, with shape `(B, 3)`.
+        translation: Translation in voxels, with shape `(B, 3)`.
```
[the remark this answers is not in the document]
Comment (fepegar) [reply src/torchio/transforms/intensity/motion.py:428]: [Generated by a coding agent]

---

Fixed in 32d60423. Aligned the docs with the long-standing behavior: both the previous and the vectorized implementation treat `translation` as voxels, normalized to `affine_grid` coordinates via `translation / (shape / 2)` (a voxel-space approximation), not millimeters. Updated the public `Motion.translation` docstring accordingly rather than changing the numerical behavior.

File: src/torchio/transforms/intensity/motion.py [hunk 3]
```diff
@@ -143,120 +151,405 @@ def _apply_motion(
...
-    t_normalized = torch.tensor(translation, dtype=torch.float32) / (shape_t / 2)
-    t_normalized = t_normalized.to(tensor.device)
+    Returns:
+        Motion-corrupted `(B, C, I, J, K)` tensor.
+    """
+    result = data.float()
+    num_segments = len(segment_parameters) + 1
+    spatial_shape = result.shape[-3:]
+    segment_size = spatial_shape[0] // num_segments
+
```
[the remark this answers is not in the document]
Comment (fepegar) [reply src/torchio/transforms/intensity/motion.py]: [Generated by a coding agent]

---

Fixed in 5fd8d687. Added a clear `ValueError` when `num_transforms + 1` exceeds the first spatial axis size (`segment_size` would be 0 and the k-space segmentation would silently degenerate to replacing the whole spectrum). Added a regression test.

File: src/torchio/transforms/intensity/blur.py [hunk 4]
```diff
@@ -111,41 +112,68 @@ def _blur_per_element(
...
         Smoothed tensor.
     """
-    if all(s <= 0 for s in sigmas):
+    sigmas_array = np.asarray(sigmas, dtype=np.float64)
+    if sigmas_array.ndim == 1:
+        sigmas_array = np.broadcast_to(sigmas_array, (data.shape[0], 3)).copy()
+    if np.all(sigmas_array <= 0):
         return data
+    return _gaussian_smooth_per_element(data, sigmas_array)
+
```
[the remark this answers is not in the document]
Comment (fepegar) [reply src/torchio/transforms/intensity/blur.py:155]: [Generated by a coding agent]

---

Fixed in 5fd8d687. `_gaussian_smooth` now routes batch-shared and uniform-sigma inputs through a single shared kernel set (`_gaussian_smooth_shared`) instead of a grouped conv with B identical kernels, restoring the cheaper path for `per_instance=False`. Verified the shared path is numerically identical to the grouped-conv path (max diff 0.0). Genuinely per-element sigmas still use the grouped `conv3d`.

File: src/torchio/transforms/spatial/anisotropy.py [hunk 5]
```diff
@@ -138,6 +128,219 @@ def apply_transform(
...
+
+    Args:
+        length: Original length along the degraded axis.
+        factors: Per-element downsampling factors.
+
+    Returns:
+        Downsampled sizes matching `round(length / factor)`.
+    """
+    sizes = torch.round(length / factors).clamp_min(1)
+    return sizes.to(torch.long)
```
[the remark this answers is not in the document]
Comment (fepegar) [reply src/torchio/transforms/spatial/anisotropy.py:223]: [Generated by a coding agent]

---

This one is actually a non-issue: in current PyTorch `torch.round` uses round-half-to-even (banker's rounding), the same rule as Python's `round`. Empirically both map `0.5->0, 1.5->2, 2.5->2, 3.5->4`, and `_downsample_sizes` divides in float64, so `length/factor` matches the scalar `original_shape[axis]/factor`. The per-instance and scalar paths therefore choose the same downsample size even on `.5` ties (e.g. `9/2 = 4.5 -> 4` in both). I added an equivalence test in 34bad8e3 with an odd length (9) and factor 2.0 that exercises exactly this tie and confirms the batched path matches the scalar path, so no rounding change is needed.

File: src/torchio/transforms/spatial/anisotropy.py [hunk 6]
```diff
@@ -138,6 +128,219 @@ def apply_transform(
...
+    factors_tensor = torch.as_tensor(
+        factors,
+        dtype=torch.float64,
+        device=data.device,
+    )
+    active = factors_tensor > 1.0
+    if not bool(active.any()):
+        return data.to(data.dtype)
+
+    output = data.clone()
```
[the remark this answers is not in the document]
Comment (fepegar) [reply src/torchio/transforms/spatial/anisotropy.py:166]: [Generated by a coding agent]

---

Fixed in 34bad8e3. `_simulate_anisotropy_per_instance` now raises a clear `ValueError` when any ACTIVE (factor > 1) per-element axis is outside `{0, 1, 2}`, matching the scalar path instead of silently no-opping. Gated-out / inactive elements (factor <= 1) with an unused axis are ignored. Added a regression test.
